tests/e2e/utils/helpers.py

```diff
@@ -0,0 +1,127 @@
+    function_arn: Optional[str]
+    function_request_id: Optional[str]
+    xray_trace_id: Optional[str]
+    extra_info: Optional[str]
+
+
+class TraceSegment(BaseModel):
+    name: str
+    metadata: Dict = {}
+    annotations: Dict = {}
+
+
+def trigger_lambda(lambda_arn: str, client: Any):
+    response = client.invoke(FunctionName=lambda_arn, InvocationType="RequestResponse")
+    return response
+
+
+@lru_cache(maxsize=10, typed=False)
+@retry(ValueError, delay=1, jitter=1, tries=10)
+def get_logs(lambda_function_name: str, log_client: Any, start_time: int, **kwargs: dict) -> List[Log]:
+    response = log_client.filter_log_events(logGroupName=f"/aws/lambda/{lambda_function_name}", startTime=start_time)
+    if not response["events"]:
+        raise ValueError("Empty response from Cloudwatch Logs. Repeating...")
+    filtered_logs = []
+    for event in response["events"]:
+        try:
+            message = Log(**json.loads(event["message"]))
+        except json.decoder.JSONDecodeError:
+            continue
+        filtered_logs.append(message)
+
+    return filtered_logs
+
+
+@lru_cache(maxsize=10, typed=False)
+@retry(ValueError, delay=1, jitter=1, tries=10)
+def get_metrics(
+    namespace: str, cw_client: Any, start_date: datetime, end_date: datetime, metric_name: str, service_name: str
+):
+    response = cw_client.get_metric_data(
+        MetricDataQueries=[
+            {
+                "Id": "m1",
+                "MetricStat": {
+                    "Metric": {
+                        "Namespace": namespace,
+                        "MetricName": metric_name,
+                        "Dimensions": [{"Name": "service", "Value": service_name}],
+                    },
+                    "Period": 600,
+                    "Stat": "Maximum",
+                },
+                "ReturnData": True,
+            },
+        ],
+        StartTime=start_date,
+        EndTime=end_date,
+    )
+    result = response["MetricDataResults"][0]
+    if not result["Values"]:
+        raise ValueError("Empty response from Cloudwatch. Repeating...")
+    return result
+
+
+@retry(ValueError, delay=1, jitter=1, tries=10)
+def get_traces(lambda_function_name: str, xray_client: Any, start_date: datetime, end_date: datetime) -> Dict:
+    paginator = xray_client.get_paginator("get_trace_summaries")
+    response_iterator = paginator.paginate(
+        StartTime=start_date,
+        EndTime=end_date,
+        TimeRangeType="Event",
+        Sampling=False,
+        FilterExpression=f'service("{lambda_function_name}")',
+    )
+
+    traces = [trace["TraceSummaries"][0]["Id"] for trace in response_iterator if trace["TraceSummaries"]]
+    if not traces:
+        raise ValueError("Empty response from X-RAY. Repeating...")
+
+    trace_details = xray_client.batch_get_traces(
+        TraceIds=traces,
+    )
+
+    return trace_details
+
+
+def find_trace_additional_info(trace: Dict) -> List[TraceSegment]:
```

Comment:
could you add a docstring? I'm not sure what the name means. I see that you're extracting metadata and annotations from a subsegment(s), so docstrings could be helpful for the untrained eye later.... and eventually even repurpose this to be generic (any origin, any subsegment name) in the future.

Reply:
Yeah, I struggle to name it correctly as I find hard to combine annotations and metadata into one meaningful word :-) Will add docstring and we can adjust the naming if you have suggestion around it